common/beerocks/bcl/include/bcl/network/socket_refactorings.h

```diff
@@ -0,0 +1,795 @@
+#ifndef socket_refactoring_H
+#define socket_refactoring_H
+
+#include <stddef.h>
+#include <stdint.h>
+
+#include <linux/if.h>
+#include <linux/if_packet.h>
+#include <linux/netlink.h>
+#include <linux/un.h>
+
+#include <algorithm>
+#include <functional>
+#include <memory>
+#include <string>
+#include <unordered_map>
+
+#include <tlvf/CmduMessageRx.h>
+#include <tlvf/common/sMacAddr.h>
+#include <tlvf/ieee_1905_1/eMessageType.h>
+
+#include <mapf/common/utils.h>
+
+#include "beerocks_event_loop.h"
+
+namespace beerocks {
+namespace sockets {
+
+/**
+ * Array of bytes used to hold data received through a socket.
+ * Code is programmed to interfaces so it does not care about which implementation is used.
+ * Unit tests can use a mock and set different expectations per test (pretend that different data
+ * has been received through the socket).
+ */
+class Buffer {
+public:
+    virtual ~Buffer()                   = default;
+    virtual const uint8_t *data() const = 0;
+    virtual size_t size() const         = 0;
+
+    uint8_t *data() { return const_cast<uint8_t *>(const_cast<const Buffer *>(this)->data()); }
+};
+
+/**
+ * First example implementation
+ */
+class BufferImpl1 : public Buffer {
+public:
+    BufferImpl1(uint8_t *data, size_t size) : m_data(data), m_size(size) {}
+    const uint8_t *data() const override { return m_data; }
+    size_t size() const override { return m_size; }
+
+private:
+    uint8_t *m_data;
+    size_t m_size;
+};
+
+/**
+ * Second example implementation
+ */
+template <size_t Size> class BufferImpl2 : public Buffer {
+public:
+    const uint8_t *data() const override { return m_data; }
+    size_t size() const override { return sizeof(m_data); }
+
+private:
+    uint8_t m_data[Size];
+};
+
+/**
+ * This interface models OS resources implementing the file descriptor interface.
+ */
+class FileDescriptor {
+public:
+    static constexpr int invalid_descriptor = -1;
+
+    virtual ~FileDescriptor() = default;
+    virtual int fd()          = 0;
+    virtual bool close()      = 0;
+};
+
+/**
+ * File descriptor implementation.
+ * This class is basically a wrapper around an `int fd`
+ * This class will be aggregated from all classes modeling OS resources implementing the file
+ * descriptor interface.
+ */
+class FileDescriptorImpl : public FileDescriptor {
+public:
+    explicit FileDescriptorImpl(int fd) : m_fd(fd)
+    {
+        // TODO: error if FileDescriptor::invalid_descriptor == fd;

```

Comment:
My TODO comment should have been "**log** error if file_descriptor == -1". Constructing with -1 **is** an error.

Reply:
Just to be clear: I think constructing a FileDescriptor with fd == -1 is an error in the sense that something has gone wrong, but not a _programming error_. In other words, I think this should be allowed:
```
FileDescriptorImpl m_fd = socket(....);
if (!m_fd)
    return false;
```
so that we can simply do the logging in the constructor.

Of course, even better would be to throw an exception :-)